ADBDEV-6986: Fix assert when using append optimized tables with GIN index

mos65o2 opened the pull request:
Fix assert when using append optimized tables with GIN index (#1221)

Using an AO table with a GIN index could cause an assert in the debug build.
The triggered assert (OffsetNumberIsValid) checks the offset (ip_posid) in the
tuple address (ItemPointerData), indirectly checking the correctness of
GinPostingList decompression. Decompression can be performed, for example,
when inserting data into an AO table with a GIN index or starting a vacuum.
The offset range is checked as for a Heap table. The correct range is 1..8192
with a BLCKSZ of 32768 bytes (see MaxOffsetNumber).
However, in AO tables, the TID offset (ip_posid) is calculated differently,
this led to the assert.

In a Heap table, ip_posid means the ItemIdData number inside the PageHeader
structure. ip_posid for Heap tables is calculated by PageGetMaxOffsetNumber()
macro, for example during insertion. ip_posid can take the value 1..8192 with
BLCKSZ size 32768 bytes.

For AO tables ip_posid is calculated as the value of the lower 16 bits of the
segfile row in AOTupleIdInit() and can take the value 1..32768. What could
cause the assert.

A larger range of ip_posid with AO tuples does not cause problems in the GIN
index, since the full tuple address (ItemPointerData) is used during
processing.

This commit weakens the offset range check (ip_posid) in the assert at the
GinPostingList decompression stage so that AO tables with GIN index can be
processed in the build with asserts.

Changes:
Only the ItemPointerIsValid() check is left;
The appendonly_with_gin_index test has been added to the parallel_schedule;
In the test query, the data volume has been increased and the search condition
has been rewritten so that the GIN index works.

KnightMurloc commented:
It looks like we already had a test for this case `appendonly_with_gin.sql`. But it doesn't work... Because there is too little data in the table. I think instead of adding a new test, it's better to refine the old one.

mos65o2 commented:
> It looks like we already had a test for this case `appendonly_with_gin.sql`. But it doesn't work... Because there is too little data in the table. I think instead of adding a new test, it's better to refine the old one.

This test does not run - it is not in the schedule list.
Move the test case from `gin.sql` to `appendonly_with_gin.sql` and include `appendonly_with_gin.sql` in the schedule?

mos65o2 committed:
Fix assert when using append optimized tables with GIN index (#1221)

Using an AO table with a GIN index could cause an assert in the debug build.
The triggered assert (OffsetNumberIsValid) checks the offset (ip_posid) in the
tuple address (ItemPointerData), indirectly checking the correctness of
GinPostingList decompression. Decompression can be performed, for example,
when inserting data into an AO table with a GIN index or starting a vacuum.
The offset range is checked as for a Heap table. The correct range is 1..8192
with a BLCKSZ of 32768 bytes (see MaxOffsetNumber).
However, in AO tables, the TID offset (ip_posid) is calculated differently,
this led to the assert.

In a Heap table, ip_posid means the ItemIdData number inside the PageHeader
structure. ip_posid for Heap tables is calculated by PageGetMaxOffsetNumber()
macro, for example during insertion. ip_posid can take the value 1..8192 with
BLCKSZ size 32768 bytes.

For AO tables ip_posid is calculated as the value of the lower 16 bits of the
segfile row in AOTupleIdInit() and can take the value 1..32768. What could
cause the assert.

A larger range of ip_posid with AO tuples does not cause problems in the GIN
index, since the full tuple address (ItemPointerData) is used during
processing.

This commit weakens the offset range check (ip_posid) in the assert at the
GinPostingList decompression stage so that AO tables with GIN index can be
processed in the build with asserts.

Also added a test.

KnightMurloc commented on src/test/regress/sql/appendonly_with_gin_index.sql:
```diff
@@ -1,19 +1,25 @@
+set enable_seqscan = 0;
+-- collect all tuples on one segment
+create temp table t1 (a text, b int default 0) with (appendonly=true) distributed by (b);
+insert into t1 select MD5(random()::text) from generate_series(1, 500000);
```
Do we really need that much data?

mos65o2 replied:
Yes, the number of lines can be reduced by 10 times. Initially, I assumed that the scheduler did not use index scanning because there was little data. And I came to this number.
Reduce to 50000?

KnightMurloc replied:
Yes, it's better this way.

KnightMurloc commented on src/test/regress/sql/appendonly_with_gin_index.sql:
```diff
@@ -1,19 +1,25 @@
+-- divide block_size by 4 because the size of ItemIdData is 4 bytes
+with query as (select a, (ctid::text::point)[1]::int as offset from t1
+              where to_tsvector('english', a) @@ to_tsquery('english', 'apple')),
+     maxHeapOffset as (select current_setting('block_size')::int/4 as value)
```
Why do we need this?

mos65o2 replied:
This gets the maximum offset number for the Heap table. Similar to [this](https://github.com/arenadata/gpdb/blob/33b6e6c8dc4afcd20e470d2b33e912d7208a3330/src/include/storage/off.h#L28). To make sure that a tuple with a large offset can be processed (not only index creation but also scanning).

KnightMurloc replied:
Add a comment about it.

mos65o2 committed:
Reduce the number of rows in the test query and add a comment

dkovalev1 commented:
Could other types of indexes be affected by this offset interpretation mismatch?

dkovalev1 commented on src/test/regress/expected/appendonly_with_gin_index.out:
```diff
@@ -1,21 +1,48 @@
+-- the maximum block size (BLCKSZ) divided by the pointer size (4 bytes), similar
+-- to the MaxOffsetNumber macro. To ensure that the AO table tuple with a large
+-- offset can be processed.
+with query as (select a, (ctid::text::point)[1]::int as offset from t1
```
Isn't ctid have format (block number, row offset)? Assuming that arrays in SQL are one-based, perhaps  we get block number as offset here?

mos65o2 replied:
Yes, ctid consists of block number and offset. The query from the test prints the offset of ctid.
You can compare the output of ctid, and the output of ctid elements:
```
postgres=# select ctid, (ctid::text::point)[0]::int as block, (ctid::text::point)[1]::int as offset from t1 limit 5;
     ctid     |  block   | offset 
--------------+----------+--------
 (33554432,2) | 33554432 |      2
 (33554432,3) | 33554432 |      3
 (33554432,4) | 33554432 |      4
 (33554432,5) | 33554432 |      5
 (33554432,6) | 33554432 |      6
```
I also compared the value of ctid[1] in the query output and `ip_posid` when processing assert. These values ​​are equivalent.

dkovalev1 commented on src/test/regress/expected/appendonly_with_gin_index.out:
```diff
@@ -1,21 +1,48 @@
+-- We define the maximum TID offset for a Heap table in the maxHeapOffset query as
+-- the maximum block size (BLCKSZ) divided by the pointer size (4 bytes), similar
+-- to the MaxOffsetNumber macro. To ensure that the AO table tuple with a large
+-- offset can be processed.
```
Could you clarify this comment?

mos65o2 replied:
We define the maximum TID offset for a Heap table in the maxHeapOffset query as the maximum block size (BLCKSZ) divided by the pointer size (4 bytes), similar to the MaxOffsetNumber macro.
But we work with the AO table.
We add enough data to the AO table so that the row we are looking for has an offset greater than the maxHeapOffset (equvalent MaxOffsetNumber).
We search for the row we are interested in using the GIN index.
Then we make sure that the row with the larger offset can be handled by the GIN index. Otherwise, we would get an error or a missing row.

dkovalev1 commented on src/test/regress/expected/appendonly_with_gin_index.out:
```diff
@@ -1,21 +1,48 @@
 (1 row)
 
+explain (costs off)
+with query as (select a, (ctid::text::point)[1]::int as offset from t1
```
What does statement plan check in the test?

mos65o2 replied:
We check that an index scan occurs, not a sequential scan (as it was in the test without the patch).

dkovalev1 commented on src/test/regress/expected/appendonly_with_gin_index.out:
```diff
@@ -1,21 +1,48 @@
-------------
- 'John'
+-- check that GIN index works with large TID offset (second ctid field) when using AO table
+-- Orca performs seq scan in this case, so disable
```
The sentence reads unfinished. Disable what?

mos65o2 replied:
...so disable Orca. I'll add this.

RekGRpth commented on src/test/regress/serial_schedule:
```diff
@@ -69,6 +69,7 @@ test: create_index_spgist
 test: create_view
 test: index_including
 test: index_including_gist
+test: appendonly_with_gin_index
```
Why run the same test twice, in parallel and sequentially?

mos65o2 replied:
I'll leave this only in `parallel_schedule`.

RekGRpth commented on src/test/regress/sql/appendonly_with_gin_index.sql:
```diff
@@ -1,19 +1,28 @@
--- Given I have an append-only table
-create table users(
```
Why was the original test content removed?

mos65o2 replied:
The original test content did not perform index scanning. And there was no check for large ctid offset.
If preferred, I can use the table and attribute names from the test rather than the ticket.

RekGRpth replied:
I think it would be better to refine the original test and write about the refinements in the description.

mos65o2 replied:
Should I keep the check that the ctid offset must be greater than what can be in the Heap table? Or edit the test with minimal changes?
``` SQL
-- Given I have an append-only table
create table users(
  first_name tsvector
) with (appendonly=true);

-- And I have a large amount of data in the table
insert into users
  select to_tsvector( md5(random()::text))
  from generate_series(1, 60000) i;

insert into users values (to_tsvector('John'));

-- When I create a GIN index on users
set gp_debug_linger=1;
CREATE INDEX users_search_idx ON users USING gin (first_name);
reset gp_debug_linger;

-- Orca performs seq scan in this case, so disable Orca.
set optimizer = 0;

-- Then I should be able to query the table
select * from users where first_name @@ to_tsquery('John');
explain select * from users where first_name @@ to_tsquery('John');

drop table users;
reset optimizer;
```

RekGRpth replied:
> edit the test with minimal changes

I think so

RekGRpth replied:
write about the refinements/changes in the description

mos65o2 commented:
> Could other types of indexes be affected by this offset interpretation mismatch?

I looked at the GIST index. It also uses TIDs with a large offset, this does not cause problems. Apparently, there was another problem, when the TID offset in the AO table was in the range 32768..65535. But this was fixed in [this](https://github.com/arenadata/gpdb/commit/c249ac7a36d9da3d25b6c419fbd07e2c9cfe954f) commit.
I did not consider other types of indexes. But by analogy with the GIN index, I can assume that the full address of the tuple is also used there, which should not cause difficulties.

mos65o2 committed:
Remove test from serial_schedule and rewrite comments

mos65o2 committed:
Use the original version of the test with corrections

mos65o2 committed:
Merge branch 'adb-7.2.0' into ADBDEV-6986

mos65o2 committed:
Modify explain output

dkovalev1 commented on src/test/regress/sql/appendonly_with_gin_index.sql:
```diff
@@ -5,15 +5,22 @@ create table users(
+insert into users values (to_tsvector('John'));
 
 -- When I create a GIN index on users
 set gp_debug_linger=1;
```
Why do we need this parameter in the test?

mos65o2 replied:
In our case, setting a delay after a fatal error makes no sense. I didn't remove it, so we agreed to fix the test with minimal changes. Should I remove this GUC?

dkovalev1 replied:
it's up to you, if it's not needed why to keep it?

mos65o2 replied:
Ok, let's delete it.

dkovalev1 commented on src/test/regress/sql/appendonly_with_gin_index.sql:
```diff
@@ -5,15 +5,22 @@ create table users(
 
```
Perhaps we can use more unique name for the table to not interfere with other tests in parallel environment?

mos65o2 replied:
This is actually a more unique name than `t` or `t1..` which are present in the parallel group tests. This test is run together with `gp_gin_index` which has different table names.

mos65o2 committed:
Remove gp_debug_linger from test

Stolb27 committed:
Merge branch 'adb-7.2.0' into ADBDEV-6986
